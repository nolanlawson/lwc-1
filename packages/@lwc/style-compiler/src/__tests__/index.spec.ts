/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
import fs from 'fs';
import path from 'path';
import { parseSync } from '@babel/core';

import { transform } from '../index';

const FIXTURES_PATH = path.resolve(__dirname, 'fixtures');

fs.readdirSync(FIXTURES_PATH).forEach((fixtureName: string) => {
    const fixtureDir = path.resolve(FIXTURES_PATH, fixtureName);
    const stats = fs.statSync(fixtureDir);

    // Ignore non directory fixtures
    if (!stats.isDirectory()) {
        return;
    }

    test(fixtureName, testFixture(fixtureDir));
});

function testFixture(fixtureDir: string) {
    const actualPath = path.resolve(fixtureDir, 'actual.css');
    const configPath = path.resolve(fixtureDir, 'config.json');

    let actualSource: string;
    if (fs.existsSync(actualPath)) {
        actualSource = fs.readFileSync(actualPath, 'utf8');
    } else {
        throw new Error(`Missing actual file ${actualPath}`);
    }

    let config: any = {};
    if (fs.existsSync(configPath)) {
        config = require(configPath);
    }

    return () => {
        let result;
        let error;

        try {
            const { code } = transform(actualSource, actualPath, config);
            result = code;
        } catch (err) {
            error = normalizeError(err);
        }

        if (error) {
            expect(error).toMatchSnapshot();
        } else {
            expect(result).toMatchSnapshot();
            // Assert that the result is valid javascript
            expect(() => parseSync(result, { babelrc: false, configFile: false })).not.toThrow();
        }
    };
}

function normalizeError(err) {
    if (err.name === 'CssSyntaxError') {
        return {
            name: err.name,
            reason: err.reason,
            column: err.column,
            line: err.line,
        };
    } else {
        return {
            name: err.name,
            message: err.message,
        };
    }
}
