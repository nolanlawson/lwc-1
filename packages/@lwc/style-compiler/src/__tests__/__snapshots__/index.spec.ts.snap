// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aria-attributes 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"[aria-labelledby]\\", shadowSelector, \\" {}[aria-labelledby=\\\\\\"bar\\\\\\"]\\", shadowSelector, \\" {}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`at-rules 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"@charset \\\\\\"utf-8\\\\\\";@namespace url(http://www.w3.org/1999/xhtml);@keyframes slidein {from {margin-left: 100%;}to {margin-left: 0%;}}@media screen and (min-width: 900px) {article\\", shadowSelector, \\" {padding: 1rem 3rem;}}@supports (display: grid) {div\\", shadowSelector, \\" {display: grid;}}@document url('https://www.example.com/') {h1\\", shadowSelector, \\" {color: green;}}@font-face {font-family: 'Open Sans';src: url('/fonts/OpenSans-Regular-webfont.woff2') format('woff2'),\\\\n        url('/fonts/OpenSans-Regular-webfont.woff') format('woff');}@viewport {width: device-width;}@counter-style thumbs {system: cyclic;symbols: '\\\\\\\\1F44D';suffix: ' ';}@font-feature-values Font One {@styleset {nice-style: 12;}}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`back-slash 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\".foo\\", shadowSelector, \\" {content: \\\\\\"\\\\\\\\\\\\\\\\\\\\\\";}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`complex-selectors 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"h1\\", shadowSelector, \\" > a\\", shadowSelector, \\" {}h1\\", shadowSelector, \\" + a\\", shadowSelector, \\" {}div.active\\", shadowSelector, \\" > p\\", shadowSelector, \\" {}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`css-import 1`] = `
Object {
  "column": 1,
  "line": 1,
  "name": "CssSyntaxError",
  "reason": "Invalid import statement, import statement only support a single parameter.",
}
`;

exports[`css-import-media-query 1`] = `
"import stylesheet0 from \\"foo\\";
import stylesheet1 from \\"./foo.css\\";

export default [stylesheet0, stylesheet1];"
`;

exports[`css-import-not-at-the-top 1`] = `
"import stylesheet0 from \\"foo\\";
import stylesheet1 from \\"./foo.css\\";

export default [stylesheet0, stylesheet1];"
`;

exports[`css-import-url 1`] = `
"import stylesheet0 from \\"foo\\";
import stylesheet1 from \\"./foo.css\\";

export default [stylesheet0, stylesheet1];"
`;

exports[`css-import-without-string 1`] = `
Object {
  "column": 1,
  "line": 1,
  "name": "CssSyntaxError",
  "reason": "Invalid import statement, unable to find imported module.",
}
`;

exports[`data-attributes 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"[data-foo]\\", shadowSelector, \\" {}[data-foo=\\\\\\"bar\\\\\\"]\\", shadowSelector, \\" {}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`deprecated-deep-selector 1`] = `
Object {
  "column": 7,
  "line": 1,
  "name": "CssSyntaxError",
  "reason": "Invalid usage of deprecated selector \\"/deep/\\".",
}
`;

exports[`deprecated-shadow-selector 1`] = `
Object {
  "column": 6,
  "line": 1,
  "name": "CssSyntaxError",
  "reason": "Invalid usage of deprecated selector \\"::shadow\\".",
}
`;

exports[`dir-attribute 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"[dir=\\\\\\"rtl\\\\\\"]\\", shadowSelector, \\" test\\", shadowSelector, \\" {order: 0;}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`dir-pseudo-class 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"[dir=\\\\\\"ltr\\\\\\"] {order: 0;}[dir=\\\\\\"ltr\\\\\\"] test\\", shadowSelector, \\" {order: 1;}[dir=\\\\\\"ltr\\\\\\"] test\\", shadowSelector, \\" {order: 2;}[dir=\\\\\\"ltr\\\\\\"] test\\", shadowSelector, \\" {order: 3;}[dir=\\\\\\"ltr\\\\\\"] test\\", shadowSelector, \\" test\\", shadowSelector, \\" {order: 4;}[dir=\\\\\\"rtl\\\\\\"] {order: 5;}[dir=\\\\\\"rtl\\\\\\"] test\\", shadowSelector, \\" {order: 6;}[dir=\\\\\\"rtl\\\\\\"] test\\", shadowSelector, \\" {order: 7;}[dir=\\\\\\"rtl\\\\\\"] test\\", shadowSelector, \\" {order: 8;}[dir=\\\\\\"rtl\\\\\\"] test\\", shadowSelector, \\" test\\", shadowSelector, \\" {order: 9;}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`dir-pseudo-class-invalid-value 1`] = `
Object {
  "column": 1,
  "line": 1,
  "name": "CssSyntaxError",
  "reason": ":dir() pseudo class expects \\"ltr\\" or \\"rtl\\" for value, but received \\"right-to-left\\".",
}
`;

exports[`escape-string-characters 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"h1\\", shadowSelector, \\":before {content: '$test\\\\\\"escaping quote(\\\\\\\\')'}h1\\", shadowSelector, \\":after {content: \\\\\\"$my test'escaping quote(\\\\\\\\\\\\\\")\\\\\\"}h2\\", shadowSelector, \\":before {content: \\\\\\"\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\"}h2\\", shadowSelector, \\":after {content: \\\\\\"\`\\\\\\"}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`global-html-attributes 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"[hidden]\\", shadowSelector, \\" {}[lang=\\\\\\"fr\\\\\\"]\\", shadowSelector, \\" {}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`grave-accents 1`] = `"export default [];"`;

exports[`host 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return (nativeShadow ? \\":host {}\\" : [hostSelector, \\" {}\\"].join(''));
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`host-functional 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [(nativeShadow ? \\":host(.foo) {}\\" : [hostSelector, \\".foo {}\\"].join('')), (nativeShadow ? [\\":host(.foo) span\\", shadowSelector, \\" {}\\"].join('') : [hostSelector, \\".foo span\\", shadowSelector, \\" {}\\"].join('')), (nativeShadow ? \\":host(:hover) {}\\" : [hostSelector, \\":hover {}\\"].join('')), (nativeShadow ? \\":host(.foo, .bar) {}\\" : [hostSelector, \\".foo,\\", hostSelector, \\".bar {}\\"].join(''))].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`id-selector 1`] = `
Object {
  "column": 2,
  "line": 1,
  "name": "CssSyntaxError",
  "reason": "Invalid usage of id selector '#foo'. Try using a class selector or some other selector.",
}
`;

exports[`id-selector-in-compound-selector 1`] = `
Object {
  "column": 2,
  "line": 1,
  "name": "CssSyntaxError",
  "reason": "Invalid usage of id selector '#foo'. Try using a class selector or some other selector.",
}
`;

exports[`invalid-css 1`] = `
Object {
  "column": 1,
  "line": 1,
  "name": "CssSyntaxError",
  "reason": "Unknown word",
}
`;

exports[`lwc-reserved-foreach-attribute-rule 1`] = `
Object {
  "column": 1,
  "line": 1,
  "name": "CssSyntaxError",
  "reason": "Invalid usage of attribute selector \\"for:each\\". \\"for:each\\" is a template directive and therefore not supported in css rules.",
}
`;

exports[`lwc-reserved-iffalse-attribute-rule 1`] = `
Object {
  "column": 1,
  "line": 1,
  "name": "CssSyntaxError",
  "reason": "Invalid usage of attribute selector \\"if:false\\". \\"if:false\\" is a template directive and therefore not supported in css rules.",
}
`;

exports[`lwc-reserved-iftrue-attribute-rule 1`] = `
Object {
  "column": 1,
  "line": 1,
  "name": "CssSyntaxError",
  "reason": "Invalid usage of attribute selector \\"if:true\\". \\"if:true\\" is a template directive and therefore not supported in css rules.",
}
`;

exports[`lwc-reserved-iterator-attribute-rule 1`] = `
Object {
  "column": 1,
  "line": 1,
  "name": "CssSyntaxError",
  "reason": "Invalid usage of attribute selector \\"iterator:iteratorName\\". \\"iterator:iteratorName\\" is a template directive and therefore not supported in css rules.",
}
`;

exports[`lwc-reserved-key-attribute-rule 1`] = `
Object {
  "column": 2,
  "line": 1,
  "name": "CssSyntaxError",
  "reason": "Invalid usage of attribute selector \\"key\\". \\"key\\" is a template directive and therefore not supported in css rules.",
}
`;

exports[`lwc-reserved-similarButAllowed-rule 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"[turkey='val']\\", shadowSelector, \\" {}[keyboard='val']\\", shadowSelector, \\" {}[notif\\\\\\\\:true='val']\\", shadowSelector, \\" {}[notfor\\\\\\\\:item='val']\\", shadowSelector, \\" {}[notfor\\\\\\\\:each='val']\\", shadowSelector, \\" {}[notiterator\\\\\\\\:name='val']\\", shadowSelector, \\" {}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`media-query 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"@media screen and (min-width: 900px) {h1\\", shadowSelector, \\" {}}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`media-query-with-host 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"@media screen and (max-width: 768px) {\\", (nativeShadow ? \\":host {width: calc(50% - 1rem);}\\" : [hostSelector, \\" {width: calc(50% - 1rem);}\\"].join('')), \\"}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`multi-line 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [(nativeShadow ? [\\":host(.selected) button\\", shadowSelector, \\" {color: #0070d2;}\\"].join('') : [hostSelector, \\".selected button\\", shadowSelector, \\" {color: #0070d2;}\\"].join('')), (nativeShadow ? [\\":host:not(.selected) button\\", shadowSelector, \\" {color: #6d6d70;transition: all .2s ease-in-out;}\\"].join('') : [hostSelector, \\":not(.selected) button\\", shadowSelector, \\" {color: #6d6d70;transition: all .2s ease-in-out;}\\"].join('')), (nativeShadow ? [\\":host:not(.selected):hover button\\", shadowSelector, \\",:host button:focus\\", shadowSelector, \\" {color: #005fb2;transform: scale(1.2);}\\"].join('') : [hostSelector, \\":not(.selected):hover button\\", shadowSelector, \\",\\", hostSelector, \\" button:focus\\", shadowSelector, \\" {color: #005fb2;transform: scale(1.2);}\\"].join('')), \\"button:focus\\", shadowSelector, \\" {outline: none;box-shadow: none;background-color: #f3f2f2;}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`multi-selectors 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"h1\\", shadowSelector, \\", h2\\", shadowSelector, \\" {}h1\\", shadowSelector, \\",h2\\", shadowSelector, \\"{}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`nester-vars-compounded 1`] = `
"import varResolver from \\"custom-properties-resolver\\";


var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\".a\\", shadowSelector, \\" {box-shadow: \\", varResolver(\\"--lwc-c-active-button-box-shadow\\",\\"0 0 2px \\" + varResolver(\\"--lwc-brand-accessible\\",\\"#0070d2\\")), \\";}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`pseudo-class-functional-selector 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\":not(p)\\", shadowSelector, \\" {}p:not(.foo, .bar)\\", shadowSelector, \\" {}:matches(ol, li, span)\\", shadowSelector, \\" {}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`pseudo-class-selector 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\":checked\\", shadowSelector, \\" {}ul\\", shadowSelector, \\" li:first-child\\", shadowSelector, \\" a\\", shadowSelector, \\" {}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`pseudo-element 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [shadowSelector, \\"::after {}h1\\", shadowSelector, \\"::before {}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`simple-selectors 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"h1\\", shadowSelector, \\" {}.foo\\", shadowSelector, \\" {}[data-foo]\\", shadowSelector, \\" {}[data-foo=\\\\\\"bar\\\\\\"]\\", shadowSelector, \\" {}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`support-query 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"@supports (display: flex) {h1\\", shadowSelector, \\" {}}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`universal-selector 1`] = `
"

var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"*\\", shadowSelector, \\" {}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`unsupported-host-context-selector 1`] = `
Object {
  "column": 1,
  "line": 1,
  "name": "CssSyntaxError",
  "reason": "Invalid usage of unsupported selector \\":host-context\\".",
}
`;

exports[`unsupported-root-selector 1`] = `
Object {
  "column": 1,
  "line": 1,
  "name": "CssSyntaxError",
  "reason": "Invalid usage of unsupported selector \\":root\\".",
}
`;

exports[`var-function 1`] = `
"import varResolver from \\"custom-properties-resolver\\";


var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"div\\", shadowSelector, \\" {color: \\", varResolver(\\"--lwc-color\\"), \\";}div\\", shadowSelector, \\" {color: \\", varResolver(\\"--lwc-color\\",\\"black\\"), \\";}div\\", shadowSelector, \\" {color: \\", varResolver(\\"--lwc-color\\"), \\" important;}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`var-function-multiple 1`] = `
"import varResolver from \\"custom-properties-resolver\\";


var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"div\\", shadowSelector, \\" {color: \\", varResolver(\\"--lwc-color\\"), \\",\\", varResolver(\\"--lwc-other\\"), \\";}div\\", shadowSelector, \\" {border: \\", varResolver(\\"--border\\",\\"1px solid rgba(0,0,0,0.1)\\"), \\";}div\\", shadowSelector, \\" {background: linear-gradient(to top,\\", varResolver(\\"--lwc-color\\"), \\",\\", varResolver(\\"--lwc-other\\"), \\");}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`var-function-nested 1`] = `
"import varResolver from \\"custom-properties-resolver\\";


var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [\\"div\\", shadowSelector, \\" {background: \\", varResolver(\\"--lwc-color\\",varResolver(\\"--lwc-other\\",\\"black\\")), \\";}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;

exports[`var-host-combined 1`] = `
"import varResolver from \\"custom-properties-resolver\\";


var cachedStylesheet

function generateCss(hostSelector, shadowSelector, nativeShadow) {
  return [(nativeShadow ? [\\":host {color: \\", varResolver(\\"--lwc-color\\"), \\";padding: 10px;}\\"].join('') : [hostSelector, \\" {color: \\", varResolver(\\"--lwc-color\\"), \\";padding: 10px;}\\"].join('')), (nativeShadow ? [\\":host(.foo) {background: linear-gradient(to top,\\", varResolver(\\"--lwc-color\\"), \\",\\", varResolver(\\"--lwc-other\\"), \\");}\\"].join('') : [hostSelector, \\".foo {background: linear-gradient(to top,\\", varResolver(\\"--lwc-color\\"), \\",\\", varResolver(\\"--lwc-other\\"), \\");}\\"].join('')), \\"div\\", shadowSelector, \\" {background: \\", varResolver(\\"--lwc-color\\",varResolver(\\"--lwc-other\\",\\"black\\")), \\";display: \\\\\\"block\\\\\\";}\\"].join('');
}

function stylesheet(hostSelector, shadowSelector, nativeShadow, hasAdoptedStyleSheets) {
  if (nativeShadow && hasAdoptedStyleSheets) {
    if (!cachedStylesheet) {
      cachedStylesheet = new CSSStyleSheet();
      cachedStylesheet.replaceSync(generateCss(hostSelector, shadowSelector, nativeShadow));
    }
    return cachedStylesheet; // fast path
  }
  return generateCss(hostSelector, shadowSelector, nativeShadow);
}
export default [stylesheet];"
`;
